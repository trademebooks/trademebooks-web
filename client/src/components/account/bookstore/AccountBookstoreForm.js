import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { MDBInput, MDBBtn, MDBIcon} from 'mdbreact'
import { toastr } from 'react-redux-toastr'
import GoogleMaps from '../LocationSearch'

const AccountBookStoreForm = ({
  auth: { user },
  getAccountSettings,
  saveAccountSettings,
  updateAuthUser
}) => {
  const [formData, setFormData] = useState({
    username: '',
    location: '',
    school: ''
  })

  const { username, location, school } = formData
  
  const handleSelect = (value) =>{
    if(value != null){
      setFormData({...formData, "location":value.description})
    }
  }

  const onChange = (e) => {
    setFormData({ ...formData, [e.target.name]: e.target.value })
  }

  const saveUsername = async (e) => {
    e.preventDefault()

    const response = await updateAuthUser({ username })

    if (response) {
      toastr.success(
        'Your username has been changed.',
        'To see your username change be taken into effect, logout then log back in.',
        { timeOut: 0 }
      )
    }
  }

  const saveSettings = async (e) => {
    e.preventDefault()

    const response = await saveAccountSettings({ location, school })

    if (response) {
      toastr.success('Your settings have been updated.')
    }
  }

  useEffect(() => {
    ;(async () => {
      const account = await getAccountSettings()

      setFormData({
        ...account,
        ...user
      })
    })()
  }, [user, getAccountSettings])

  return (
    <>
      <h3 className="mb-4 font-weight-bold">Bookstore Settings</h3>

      <div className="mt-5">
        <h4>Change Username</h4>
        <MDBInput
          label="Username"
          icon="user"
          group
          type="text"
          name="username"
          onChange={onChange}
          required
          value={username}
        />
        <MDBBtn onClick={saveUsername}>Save Changes</MDBBtn>
      </div>

      <div className="mt-5">
        <h4>Location</h4>
        <div className="d-flex flex-row">
          <MDBIcon size="2x" icon = "map" className="mt-3 pr-2" />
          <GoogleMaps handleChange={handleSelect} placeholder={formData.location} className="w-100"/>
        </div>
      </div>

      <div>
        <h4>School</h4>
          <MDBInput
            label="Current School"
            icon="school"
            group
            type="text"
            name="school"
            onChange={onChange}
            minLength="6"
            validate
            value={school}
          />
      </div>

      <div>
        <MDBBtn onClick={saveSettings}>Save Changes</MDBBtn>
      </div>
    </>
  )
}

AccountBookStoreForm.propTypes = {
  auth: PropTypes.object.isRequired,
  getAccountSettings: PropTypes.func.isRequired,
  saveAccountSettings: PropTypes.func.isRequired,
  updateAuthUser: PropTypes.func.isRequired
}

export default AccountBookStoreForm
