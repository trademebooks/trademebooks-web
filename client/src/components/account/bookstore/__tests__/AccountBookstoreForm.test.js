import React from 'react'
import renderer from 'react-test-renderer'
import { render, act, fireEvent, waitFor, screen } from '@testing-library/react'

import AccountBookstoreForm from '../AccountBookstoreForm'

describe('Component: AccountBookstoreForm', () => {
  const promise = Promise.resolve()

  const props = {
    auth: {
      user: {
        first_name: 'john',
        last_name: 'doe'
      }
    },
    getAccountSettings: jest.fn(() => {
      return promise
    }),
    saveAccountSettings: jest.fn(() => {
      return promise
    }),
    updateAuthUser: jest.fn(() => {
      return promise
    })
  }

  it('should match the snapshot', async () => {
    const accountBookstoreFormSnapshot = renderer
      .create(<AccountBookstoreForm {...props} />)
      .toJSON()

    expect(accountBookstoreFormSnapshot).toMatchSnapshot()
  })

  it('should render properly', async () => {
    const { getAllByRole, getByText, getAllByText } = render(
      <AccountBookstoreForm {...props} />
    )

    getByText('Bookstore Settings')
    getByText('Change Username')
    getByText('Location')
    getByText('School')

    getAllByRole('button')
    getAllByText('Save Changes')

    await act(() => {
      return promise
    })
  })

  // fireEvent
  it('should save the changes - buttons 1 and 2', async () => {
    const { getAllByText } = render(<AccountBookstoreForm {...props} />)

    const saveChangesButton1 = getAllByText('Save Changes')[0]
    const saveChangesButton2 = getAllByText('Save Changes')[1]

    fireEvent.click(saveChangesButton1)
    fireEvent.click(saveChangesButton2)

    await act(() => {
      return promise
    })
  })
})
