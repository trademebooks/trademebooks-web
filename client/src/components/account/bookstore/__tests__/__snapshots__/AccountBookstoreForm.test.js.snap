// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AccountBookstoreForm should match the snapshot 1`] = `
Array [
  <h3
    className="mb-4 font-weight-bold"
  >
    Bookstore Settings
  </h3>,
  <div
    className="mt-5"
  >
    <h4>
      Change Username
    </h4>
    <div
      className="md-form form-group"
    >
      <i
        className="fa fa-user prefix"
        data-test="fa"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      />
      <input
        aria-disabled={false}
        className="form-control"
        data-test="input"
        disabled={false}
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        required={true}
        type="text"
        value=""
      />
      <label
        aria-labelledby=""
        className=""
        data-error=""
        data-success=""
        id=""
        onClick={[Function]}
      >
        Username
      </label>
    </div>
    <button
      className="btn-default btn Ripple-parent"
      data-test="button"
      onClick={[Function]}
      onMouseUp={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      Save Changes
      <div
        className="Ripple "
        data-test="waves"
        style={
          Object {
            "height": "0px",
            "left": "0px",
            "top": "0px",
            "width": "0px",
          }
        }
      />
    </button>
  </div>,
  <div
    className="mt-5"
  >
    <h4>
      Location
    </h4>
    <div
      className="d-flex flex-row"
    >
      <i
        className="fa fa-map fa-2x mt-3 pr-2"
        data-test="fa"
      />
      <div
        aria-expanded={false}
        aria-owns={null}
        className="MuiAutocomplete-root py-2 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink={false}
            htmlFor="google-map-demo"
            id="google-map-demo-label"
          >
            Add a Location
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="both"
              aria-controls={null}
              aria-invalid={false}
              autoCapitalize="none"
              autoComplete="off"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              disabled={false}
              id="google-map-demo"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              required={false}
              spellCheck="false"
              type="text"
              value=""
            />
            <div
              className="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Clear"
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
              </button>
              <button
                aria-label="Open"
                className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Open"
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-4"
              >
                <span>
                  Add a Location
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div>
    <h4>
      School
    </h4>
    <div
      className="row pr-0"
      data-test="row"
    >
      <div
        className="col-1 mt-3 w-75 px-0"
        data-test="col"
      >
        <i
          className="fa fa-school fa-2x ml-3"
          data-test="fa"
        />
      </div>
      <div
        className="col-11 px-0"
        data-test="col"
      >
        <div
          aria-expanded={false}
          aria-owns={null}
          className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
          name="school"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="combobox"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink={false}
              htmlFor="my-school"
              id="my-school-label"
            >
              Select School
            </label>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
              onClick={[Function]}
            >
              <input
                aria-activedescendant={null}
                aria-autocomplete="both"
                aria-controls={null}
                aria-invalid={false}
                autoCapitalize="none"
                autoComplete="off"
                autoFocus={false}
                className="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
                disabled={false}
                id="my-school"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                required={false}
                spellCheck="false"
                type="text"
                value=""
              />
              <div
                className="MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Clear"
                  className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  title="Clear"
                  type="button"
                >
                  <span
                    className="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                </button>
                <button
                  aria-label="Open"
                  className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  title="Open"
                  type="button"
                >
                  <span
                    className="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div>
    <button
      className="btn-default btn Ripple-parent"
      data-test="button"
      onClick={[Function]}
      onMouseUp={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      Save Changes
      <div
        className="Ripple "
        data-test="waves"
        style={
          Object {
            "height": "0px",
            "left": "0px",
            "top": "0px",
            "width": "0px",
          }
        }
      />
    </button>
  </div>,
]
`;
